### ISSUE TITLE
[vim9script] No way to type a variable/argument as an object

### ISSUE THREAD

**Issue by errael** [vim/vim#13120]
### Steps to reproduce

Using `o._G()` in `C.S(o: any)` in the following get's the familiar runtime error
```
E715: Dictionary required
```

```
vim9script

class C
    def _G()
        echo "C._G()"
    enddef
    static def S(o_any: any)    # Can't do "o: C" (known issue) or "o: object"
        o_any._G()
    enddef
endclass

var o1 = C.new()
C.S(o1)
```

### Expected behaviour

If `C.S()` could be defined like (note the `: object`)
```
    static def S(o_obj: object)
        o_obj._G()
    enddef
```
then there would be no ambiguity, and it could be processed correctly, requiring a runtime check.
Note: there's already a `v:t_object`.

It's possible that there's a plan to generate an instruction DICT_OR_CLASS_REFERENCE. But using `v: object` seems to be more like vim9scrip and the type `any` defaults to dict during parsing if the type is unclear from context, having a type object might be less prone to erroneous corner cases than trying to disambiguate, and I think it could be used in other situations. For example, currently you can produce
```
object<Unknown>
```
with
```
var o = null_object
typename(o)
```
but doing
```
var o: object
```
Is more intuitive.

BTW. a workaround for the original problem is
```
    static def S(o_any: any)
        var o_typed: C = o_any
        o_typed._G()
    enddef
```
or if you want to get fancy
```
    static def S(o_any: any)
        if o_any->instanceof(C)
            var o_typed: C = o_any
            o_typed._G()
        else
            throw "C.S() argument must be of type C"
        endif
    enddef
```

### Version of Vim

9.0.1909

### Environment

ubuntu

**Comment by yegappan** [comment]
It looks like this issue is related to the following item in the todo list:

```
"obj.Method()" does not always work in a compiled function, assumes "obj"
is a dictionary.  #12196  Issue #12024 might be the same problem.
Issue #11822: any.Func() can be a dict or an object call, need to handle
this at runtime.  Also see #12198 for an example.
Possibly issue #11981 can be fixed at the same time (has two examples).
```

I spent some time looking into this issue.  As you mentioned above, the fix for
this issue will need significant changes.

**Comment by errael** [comment]
> It looks like this issue is related to the following item in the todo list:

Yes. It seems like it might be useful, in general, to be able to do `var o: object`. It might be relatively simple to implement typing something as an `object`, then there are far fewer ambiguous cases. But maybe typing something as an object is just a hack.

**Comment by yegappan** [comment]
This issue is resolved now.  The type of `o_any` in the example above can be changed to `C`:

    static def S(o_any: C)
        o_any._G()
    enddef

The example also work with `any` type.
